### PR TITLE
perf: upgrade redis deps to enable compression

### PR DESCRIPTION
**Pull request by dcherman**
Newer versions of the redis go libraries include built-in support for
compression of values going to and from redis.  Since the controller
is extremely chatty and makes heavy use of redis for caching, this
reduces the amount of bandwidth requires significantly.

This change should be backwards compatible since the redis libraries
detect whether or not compression was used and does not error in the
no compression case.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

**Comment by dcherman** [comment]
For a relatively small installation of ArgoCD, here's the reduction in network traffic that I observed.  I'll let you determine exactly when I released this change to my cluster :). Everything seems to be working, including migrating from an old installation to a new one.  Note that since the existing code does *not* support compression, a rollback of this version would fail to read the cached entries.

image
image
image

File: util/cache/redis.go [hunk 1]
```diff
@@ -5,59 +5,52 @@ import (
...
+	return r.cache.Set(&rediscache.Item{
+		Key:   item.Key,
+		Value: item.Object,
+		TTL:   expiration,
 	})
 }
 
 func (r *redisCache) Get(key string, obj interface{}) error {
-	err := r.codec.Get(key, obj)
+	err := r.cache.Get(context.TODO(), key, obj)
```
Comment (dcherman) [inline util/cache/redis.go:40]: We might want to propagate contexts to these functions in the future if available, however I used `context.TODO` to keep this PR focused/minimal

File: util/cache/redis.go [hunk 2]
```diff
@@ -74,24 +67,44 @@ func (r *redisCache) OnUpdated(ctx context.Context, key string, callback func()
...
-		return func(cmd redis.Cmder) error {
-			startTime := time.Now()
-			err := oldProcess(cmd)
-			registry.IncRedisRequest(err != nil && err != redis.Nil)
-			duration := time.Since(startTime)
-			registry.ObserveRedisRequestDuration(duration)
-			return err
-		}
-	})
+	client.AddHook(&redisHook{registry: registry})
```
Comment (dcherman) [inline util/cache/redis.go:109]: Should probably use a pointer here

**Comment by dcherman** [comment]
@jessesuen In #4256 you mentioned that we should make this configurable for perf reasons on local redis, but I'm not sure it makes sense to make that configurable based on this chart.  If you have a few mins, it'd be awesome if you can pull down this PR to see if you can replicate my results since this is pretty significant.

image

**Comment by jessesuen** [comment]
If this has benefits across the board for both local and managed versions of redis, then yes I agree that there should not be a knob for this.

**Review by jessesuen — APPROVED** [review]

File: util/cache/redis.go (hunk 2 above)
Comment (dcherman) [reply util/cache/redis.go:109]: No need, this is an interface.

**Pull request merged**